process/block/preprocess/transactionsV2.go

```diff
@@ -0,0 +1,416 @@
+		txs.accountsInfo[string(tx.GetSndAddr())] = &txShardInfo{senderShardID: senderShardID, receiverShardID: receiverShardID}
+		txs.mutAccountsInfo.Unlock()
+
+		if err != nil && !errors.Is(err, process.ErrFailedTransaction) {
+			if errors.Is(err, process.ErrHigherNonceInTransaction) {
+				senderAddressToSkip = tx.GetSndAddr()
+			}
+
+			numTxsBad++
+			log.Trace("bad tx",
+				"error", err.Error(),
+				"hash", txHash,
+			)
+
+			err = txs.accounts.RevertToSnapshot(snapshot)
+			if err != nil {
+				log.Warn("revert to snapshot", "error", err.Error())
+			}
+
+			txs.gasHandler.RemoveGasConsumed([][]byte{txHash})
+			txs.gasHandler.RemoveGasRefunded([][]byte{txHash})
+
+			gasConsumedByMiniBlocksInSenderShard = oldGasConsumedByMiniBlocksInSenderShard
+			mapGasConsumedByMiniBlockInReceiverShard[receiverShardID] = oldGasConsumedByMiniBlockInReceiverShard
+			totalGasConsumedInSelfShard = oldTotalGasConsumedInSelfShard
+			continue
+		}
+
+		senderAddressToSkip = []byte("")
+
+		gasRefunded := txs.gasHandler.GasRefunded(txHash)
+		mapGasConsumedByMiniBlockInReceiverShard[receiverShardID] -= gasRefunded
+		if senderShardID == receiverShardID {
+			gasConsumedByMiniBlocksInSenderShard -= gasRefunded
+			totalGasConsumedInSelfShard -= gasRefunded
+		}
+
+		if errors.Is(err, process.ErrFailedTransaction) {
+			if !firstInvalidTxFound {
+				firstInvalidTxFound = true
+				txs.blockSizeComputation.AddNumMiniBlocks(1)
+			}
+
+			txs.blockSizeComputation.AddNumTxs(1)
+			numTxsFailed++
+			continue
+		}
+
+		if isAddressSet {
+			ok = txs.balanceComputation.SubBalanceFromAddress(tx.GetSndAddr(), txMaxTotalCost)
+			if !ok {
+				log.Error("createAndProcessMiniBlocksFromMeV2.SubBalanceFromAddress",
+					"sender address", tx.GetSndAddr(),
+					"tx max total cost", txMaxTotalCost,
+					"err", process.ErrInsufficientFunds)
+			}
+		}
+
+		if len(miniBlock.TxHashes) == 0 || firstMiniBlockSplitForReceiverShardFound {
+			txs.blockSizeComputation.AddNumMiniBlocks(1)
+		}
+
+		if isReceiverSmartContractAddress {
+			mapScsForShard[receiverShardID]++
+			mapSCTxs[string(txHash)] = struct{}{}
+		} else {
+			mapTxsForShard[receiverShardID]++
+		}
+
+		miniBlock.TxHashes = append(miniBlock.TxHashes, txHash)
+		txs.blockSizeComputation.AddNumTxs(1)
+		if isCrossShardScCallOrSpecialTx {
+			if !firstCrossShardScCallOrSpecialTxFound {
+				firstCrossShardScCallOrSpecialTxFound = true
+				txs.blockSizeComputation.AddNumMiniBlocks(1)
+			}
+			//we need to increment this as to account for the corresponding SCR hash
+			txs.blockSizeComputation.AddNumTxs(core.AdditionalScrForEachScCallOrSpecialTx)
+			numCrossShardScCallsOrSpecialTxs++
+		}
+		numTxsAdded++
+	}
+
+	miniBlocks := txs.getMiniBlockSliceFromMapV2(mapMiniBlocks, mapSCTxs)
+	scheduledMiniBlocks := txs.createScheduledMiniBlocks(haveTime, isShardStuck, isMaxBlockSizeReached, sortedTxs, mapSCTxs)
+	miniBlocks = append(miniBlocks, scheduledMiniBlocks...)
+
+	log.Debug("createAndProcessMiniBlocksFromMeV2",
```

Comment:
you may separate these write/log lines into a separate function.

Reply:
Done